src/dds/usage_reqs.test.cpp

```diff
@@ -0,0 +1,259 @@
+#include "./usage_reqs.hpp"
+
+#include <fmt/core.h>
+#include <fmt/ranges.h>
+#include <range/v3/view/transform.hpp>
+
+#include <algorithm>
+
+#include <catch2/catch.hpp>
+
+using Catch::Matchers::Contains;
+
+namespace {
+class IsRotation : public Catch::MatcherBase<std::vector<lm::usage>> {
+    std::vector<lm::usage> cycle;
+
+public:
+    IsRotation(std::vector<lm::usage> cycle)
+        : cycle(std::move(cycle)) {}
+
+    bool match(const std::vector<lm::usage>& other) const override {
+        if (cycle.size() != other.size())
+            return false;
+        if (cycle.empty())
+            return true;
+
+        std::vector<lm::usage> other_cpy(other);
+
+        const auto eq = [](const lm::usage& lhs, const lm::usage& rhs) {
+            return lhs.namespace_ == rhs.namespace_ && lhs.name == rhs.name;
+        };
+
+        auto cycle_start = std::find_if(other_cpy.begin(),
+                                        other_cpy.end(),
+                                        [&](const lm::usage& it) { return eq(it, cycle.front()); });
+        std::rotate(other_cpy.begin(), cycle_start, other_cpy.end());
+
+        return std::equal(cycle.begin(), cycle.end(), other_cpy.begin(), eq);
+    }
+
+    std::string describe() const override {
+        return fmt::format("is a rotation of: {}",
+                           fmt::join(cycle | ranges::views::transform([](const lm::usage& usage) {
+                                         return fmt::format("'{}/{}'",
+                                                            usage.namespace_,
+                                                            usage.name);
+                                     }),
+                                     ", "));
+    }
+};
+}  // namespace
+
+TEST_CASE("Cyclic dependencies are rejected") {
+    dds::usage_requirement_map reqs;
+    {
+        std::vector<lm::usage> previous;
+        for (int i = 0; i < 10; ++i) {
+            std::string dep_name = fmt::format("dep{}", i + 10);
+
+            reqs.add(dep_name,
+                     dep_name,
+                     lm::library{
+                         /* name = */ dep_name,
+                         /* linkable_path = */ std::nullopt,
+                         /* include_paths = */ {},
+                         /* preproc_defs = */ {},
+                         /* uses = */ previous,
+                         /* links = */ {},
+                         /* special_uses */ {},
+                     });
```

Comment:
Designated initializers should be available for `lm::library`. You can also go ahead and add NSDMIs for the members to prevent the "uninitialized-members" warnings. That could make these much more terse.